### PR TITLE
Improve consistency and accessibility of tooltips and popovers

### PR DESCRIPTION
**Pull request by nwalters512**
While working on #11429, I noticed some accessibility rough edges and inconsistencies with the way our tooltips and popovers work. This PR attempts to improve things.

Closes #11443.

### CODE REVIEW

File: apps/prairielearn/assets/scripts/instructorAssessmentManualGradingAssessmentQuestionClient.ts [hunk 1]
```diff
@@ -155,20 +155,24 @@ onDocumentReady(() => {
                 ? html`
                     <a
                       href="#"
-                      class="badge rounded-pill text-bg-danger"
-                      title="Instance question has ${row.open_issue_count} open ${row.open_issue_count >
+                      class="badge badge-pill badge-danger"
```
Comment (nwalters512) [inline apps/prairielearn/assets/scripts/instructorAssessmentManualGradingAssessmentQuestionClient.ts:158]:
```suggestion
                      class="badge rounded-pill text-bg-danger"
```

File: apps/prairielearn/assets/scripts/lib/qrCodeButton.ts
Comment (nwalters512) [inline apps/prairielearn/assets/scripts/lib/qrCodeButton.ts:1]: TODO: remove this file.

File: apps/prairielearn/src/components/ChangeIdButton.html.ts [hunk 2]
```diff
@@ -25,13 +25,12 @@ export function ChangeIdButton({
...
       data-other-values="${JSON.stringify(otherValues)}"
-      title="Change ${label}"
+      data-bs-title="Change ${label}"
       data-bs-content="${escapeHtml(
         ChangeIdForm({ id, label, currentValue, extraHelpText, csrfToken, action }),
       )}"
     >
-      <i class="fa fa-i-cursor"></i>
-      <span>Change ${label}</span>
+      <i class="fa fa-i-cursor"></i> Change ${label}
```
Comment (nwalters512) [inline apps/prairielearn/src/components/ChangeIdButton.html.ts:33]:
```suggestion
      <i class="fa fa-i-cursor"></i>
      <span>Change ${label}</span>
```

File: apps/prairielearn/src/components/FileBrowser.html.ts [hunk 3]
```diff
@@ -418,9 +418,8 @@ function FileBrowserActions({
             data-bs-container="body"
             data-bs-html="true"
             data-bs-placement="auto"
-            title="Upload file"
+            data-bs-title="Upload file"
             data-bs-content="${escapeHtml(FileUploadForm({ file: fileInfo, csrfToken }))}"
-            data-bs-trigger="click"
```
Comment (nwalters512) [inline apps/prairielearn/src/components/FileBrowser.html.ts:423]: This is the default; it's not needed.

File: apps/prairielearn/src/components/SyncProblemButton.html.ts [hunk 4]
```diff
@@ -18,14 +18,13 @@ ${unsafeHtml(ansiUp.ansi_to_html(output))}</pre
 
   return html`
     <button
-      class="btn btn-xs btn-ghost me-1"
+      class="btn btn-xs btn-ghost mr-1"
```
Comment (nwalters512) [inline apps/prairielearn/src/components/SyncProblemButton.html.ts:21]:
```suggestion
      class="btn btn-xs btn-ghost me-1"
```

File: apps/prairielearn/src/pages/instructorAssessmentManualGrading/assessment/assessment.html.ts [hunk 5]
```diff
@@ -145,8 +145,8 @@ function AssessmentQuestionRow({
                 href="#"
                 class="ms-2 text-info"
                 data-bs-toggle="tooltip"
-                data-boundary="window"
-                title="This question uses a rubric"
+                data-bs-boundary="window"
```
Comment (nwalters512) [inline apps/prairielearn/src/pages/instructorAssessmentManualGrading/assessment/assessment.html.ts:148]: TODO: make this change everywhere in another PR.

File: apps/prairielearn/src/pages/instructorAssessmentSettings/instructorAssessmentSettings.html.ts [hunk 6]
```diff
@@ -180,6 +181,11 @@ export function InstructorAssessmentSettings({
                 The link that students will use to access this assessment.
               </small>
             </div>
+            ${QRCodeModal({
+              id: 'student-link-qr-code-modal',
```
Comment (nwalters512) [inline apps/prairielearn/src/pages/instructorAssessmentSettings/instructorAssessmentSettings.html.ts:185]: TODO: check if this matches our conventions.

File: apps/prairielearn/src/pages/instructorCourseAdminSharing/instructorCourseAdminSharing.html.ts [hunk 7]
```diff
@@ -165,10 +165,9 @@ export function InstructorCourseAdminSharing({
                       <button
                         type="button"
                         class="btn btn-xs btn-secondary mx-2"
-                        id="chooseSharingName"
```
Comment (nwalters512) [inline apps/prairielearn/src/pages/instructorCourseAdminSharing/instructorCourseAdminSharing.html.ts:168]: AFAICT this `id` attribute was unused.

File: apps/prairielearn/src/pages/instructorInstanceAdminLti/instructorInstanceAdminLti.html.ts [hunk 8]
```diff
@@ -140,27 +140,36 @@ function LtiCredentialsCard({
                   <tr>
                     <td>
                       <code>${config.ltiRedirectUrl}</code>
-                      <i
-                        class="far fa-copy"
-                        title="Copy to clipboard"
+                      <button
+                        type="button"
+                        class="btn btn-sm btn-ghost"
+                        aria-label="Copy redirect URL to clipboard"
```
Comment (nwalters512) [inline apps/prairielearn/src/pages/instructorInstanceAdminLti/instructorInstanceAdminLti.html.ts:146]: Minor accessibility improvement. LTI 1.1 isn't maintained but this will help us if it comes up on any future accessibility reviews.

File: apps/prairielearn/src/pages/instructorRequestCourse/instructorRequestCourse.html.ts [hunk 9]
```diff
@@ -200,7 +200,6 @@ function CourseNewRequestCard({ csrfToken }: { csrfToken: string }): HtmlValue {
               id="cr-shortname"
               placeholder="MATH 101"
               pattern="[a-zA-Z]+ [a-zA-Z0-9]+"
-              title="this is a series of letters, followed by a space, followed by a series of numbers and/or letters"
```
Comment (nwalters512) [inline apps/prairielearn/src/pages/instructorRequestCourse/instructorRequestCourse.html.ts:203]: A `title` on an input isn't useful.

File: apps/prairielearn/src/pages/instructorAssessmentSettings/instructorAssessmentSettings.html.ts (hunk 6 above)
Comment (nwalters512) [reply apps/prairielearn/src/pages/instructorAssessmentSettings/instructorAssessmentSettings.html.ts]: Changed in b67ca6c.

File: apps/prairielearn/src/pages/instructorAssessmentManualGrading/assessment/assessment.html.ts (hunk 5 above)
Comment (nwalters512) [reply apps/prairielearn/src/pages/instructorAssessmentManualGrading/assessment/assessment.html.ts:148]: Turns out we can't make this change yet.

```suggestion
                data-boundary="window"
```

TODO in another PR.

File: apps/prairielearn/assets/scripts/instructorAssessmentInstancesClient.tsx [hunk 10]
```diff
@@ -347,7 +347,7 @@ onDocumentReady(() => {
           <button
             class="btn btn-xs btn-ghost"
             type="button"
-            title="Show remaining time help"
+            aria-label="Remaining time help"
             data-toggle="modal"
             data-target="#time-remaining-help"
           >
```
[the remark this answers is not in the document]
Comment (nwalters512) [reply apps/prairielearn/assets/scripts/instructorAssessmentInstancesClient.tsx]: This was fixed with a rebase.

File: apps/prairielearn/assets/scripts/instructorAssessmentInstancesClient.tsx [hunk 11]
```diff
@@ -285,7 +285,7 @@ onDocumentReady(() => {
                 <button
                   class="btn btn-xs btn-ghost"
                   type="button"
-                  title="Show roles help"
+                  aria-label="Roles help"
                   data-toggle="modal"
                   data-target="#role-help"
                 >
```
[the remark this answers is not in the document]
Comment (nwalters512) [reply apps/prairielearn/assets/scripts/instructorAssessmentInstancesClient.tsx]: This was fixed with a rebase.

File: apps/prairielearn/assets/scripts/instructorAssessmentInstancesClient.tsx [hunk 12]
```diff
@@ -328,7 +328,7 @@ onDocumentReady(() => {
           <button
             class="btn btn-xs btn-ghost"
             type="button"
-            title="Show duration help"
+            aria-label="Duration help"
             data-toggle="modal"
             data-target="#duration-help"
           >
```
[the remark this answers is not in the document]
Comment (nwalters512) [reply apps/prairielearn/assets/scripts/instructorAssessmentInstancesClient.tsx]: This was fixed with a rebase.

File: apps/prairielearn/assets/scripts/instructorAssessmentInstancesClient.tsx [hunk 13]
```diff
@@ -378,7 +378,7 @@ onDocumentReady(() => {
           <button
             class="btn btn-xs btn-ghost"
             type="button"
-            title="Show fingerprint changes help"
+            aria-label="Fingerprint changes help"
             data-toggle="modal"
             data-target="#fingerprint-changes-help"
           >
```
[the remark this answers is not in the document]
Comment (nwalters512) [reply apps/prairielearn/assets/scripts/instructorAssessmentInstancesClient.tsx]: This was fixed with a rebase.

File: apps/prairielearn/assets/scripts/lib/qrCodeButton.ts [hunk 14]
```diff
@@ -3,21 +3,29 @@
...
-      }
+      if (!content) return;
+
+      const qrCodeSvg = new QR({ content, container: 'svg-viewbox' }).svg();
+
+      const modal = Modal({
+        id: 'qr-code-modal',
+        title: 'QR Code',
+        body: qrCodeSvg,
+      });
```
[the remark this answers is not in the document]
Comment (nwalters512) [reply apps/prairielearn/assets/scripts/lib/qrCodeButton.ts]: This whole file was since removed.

File: apps/prairielearn/src/components/CourseRequestsTable.html.ts [hunk 15]
```diff
@@ -87,10 +87,9 @@ export function CourseRequestsTable({
                             class="btn btn-sm btn-danger text-nowrap me-2"
                             data-bs-toggle="popover"
                             data-bs-container="body"
-                            data-boundary="window"
```
Comment (nwalters512) [inline apps/prairielearn/src/components/CourseRequestsTable.html.ts:90]: AFAICT this isn't necessary for tooltips/popovers when we set `data-bs-container="body"`. I wasn't able to reproduce any weird overflow or clipping issues even without this attribute.

File: apps/prairielearn/assets/scripts/behaviors/dropdown.ts [hunk 16]
```diff
@@ -19,20 +38,32 @@ onDocumentReady(() => {
...
-          // a `data-bs-toggles-popover` attribute.
+          // a `data-bs-toggle-popover` attribute.
           if (clickEvent?.target.closest('[data-bs-toggle-popover]')) {
             event.preventDefault();
             return;
           }
 
           // If the click occurred inside a popover, prevent the dropdown from hiding.
+          // TODO: handle focus.
+          console.log(event);
```
Comment (nwalters512) [inline apps/prairielearn/assets/scripts/behaviors/dropdown.ts:49]:
```suggestion
```

[the remark this answers is not in the document]
Comment (nwalters512) [reply apps/prairielearn/assets/scripts/behaviors/dropdown.ts]: Good catch, I removed these in e9ea56e.

File: apps/prairielearn/assets/scripts/behaviors/bootstrap-compat.ts
Comment (nwalters512) [inline apps/prairielearn/assets/scripts/behaviors/bootstrap-compat.ts:1]: This lives on in the dropdown behavior. It's not longer just part of this compat layer.

File: apps/prairielearn/assets/scripts/behaviors/dropdown.ts [hunk 17]
```diff
@@ -31,8 +50,18 @@ onDocumentReady(() => {
             return;
           }
 
+          // If the current document focus is inside a popover, prevent the dropdown from hiding.
+          if (document.activeElement?.closest('.popover')) {
+            event.preventDefault();
+            return;
+          }
```
Comment (nwalters512) [inline apps/prairielearn/assets/scripts/behaviors/dropdown.ts:57]: This is an enhancement to the behavior on the "actions" dropdown on individual rows on the "students" tab. Without this, it was impossible to tab into the popover.

File: apps/prairielearn/assets/scripts/behaviors/tooltip.ts [hunk 18]
```diff
@@ -3,9 +3,51 @@ import { observe } from 'selector-observer';
...
+  observe('[data-bs-toggle~="tooltip"], [data-bs-toggle-tooltip="true"]', {
+    constructor: HTMLElement,
     add(el) {
       new window.bootstrap.Tooltip(el);
+
+      // Bootstrap doesn't support a single element triggering multiple things.
+      // There are cases where we want this behavior, e.g. to have a tooltip
+      // label on a button that opens a modal. We achieve that by looking for
+      // things like `data-bs-toggle="modal tooltip"` and stripping the `tooltip`
+      // piece out after initializing the tooltip.
```
Comment (nwalters512) [inline apps/prairielearn/assets/scripts/behaviors/tooltip.ts:15]: Is this too clever? I've wanted this a few times now. This is now used for the help button on the issues page.

File: apps/prairielearn/assets/scripts/instructorAssessmentInstancesClient.tsx [hunk 19]
```diff
@@ -247,7 +247,7 @@ onDocumentReady(() => {
                 <button
                   class="btn btn-xs btn-ghost"
                   type="button"
-                  title="Show roles help"
+                  aria-label="Roles help"
```
Comment (nwalters512) [inline apps/prairielearn/assets/scripts/instructorAssessmentInstancesClient.tsx:250]: IMO it makes sense to avoid using words like "Show" for these labels. Thoughts?

File: apps/prairielearn/assets/scripts/lib/qrCodeButton.ts
Comment (nwalters512) [inline apps/prairielearn/assets/scripts/lib/qrCodeButton.ts:1]: I replaced the popover with a modal that we could render on the server. This avoids extra code on the client.

File: apps/prairielearn/src/pages/instructorAssessmentGroups/instructorAssessmentGroups.html.ts [hunk 20]
```diff
@@ -204,32 +204,39 @@ export function InstructorAssessmentGroups({
...
-                                            title="Remove members"
+                                    ${row.users.length > 0
+                                      ? html`
+                                          <button
+                                            class="dropdown-item js-group-action"
+                                            data-bs-toggle="popover"
+                                            data-bs-container="body"
+                                            data-bs-html="true"
+                                            data-bs-placement="auto"
+                                            data-bs-title="Remove members"
```
Comment (nwalters512) [inline apps/prairielearn/src/pages/instructorAssessmentGroups/instructorAssessmentGroups.html.ts:215]: I changed the markup here to allow `html` to always contain a complete HTML element, which is better for syntax highlighting and formatting. Dynamic attributes like this is the kind of thing that JSX is much more suited for.

File: apps/prairielearn/src/pages/instructorCourseAdminSharing/instructorCourseAdminSharing.html.ts [hunk 21]
```diff
@@ -165,10 +165,9 @@ export function InstructorCourseAdminSharing({
...
-                        title="Choose Sharing Name"
-                        data-bs-toggle="modal"
-                        data-bs-target="#chooseSharingNameModal"
+                        aria-label="Choose Sharing Name"
```
Comment (nwalters512) [inline apps/prairielearn/src/pages/instructorCourseAdminSharing/instructorCourseAdminSharing.html.ts:168]: A label is needed because the text content here is hidden on narrow viewports. The same goes for some buttons on the staff page.

File: apps/prairielearn/src/tests/accessibility/index.test.ts [hunk 22]
```diff
@@ -38,6 +38,24 @@ async function loadPageJsdom(url: string): Promise<JSDOM> {
...
 async function checkPage(url: string) {
   const page = await loadPageJsdom(SITE_URL + url);
+
+  // We need to mimic what Bootstrap will do at runtime for tooltips and popovers:
+  // it will copy the `data-bs-title` attribute to the `aria-label` attribute if
+  // the element doesn't already have an `aria-label` attribute and if its text
+  // content is empty.
+  //
+  // Without this, we'll get a bunch of false positives that don't reflect
+  // the actual behavior at runtime.
```
Comment (nwalters512) [inline apps/prairielearn/src/tests/accessibility/index.test.ts:48]: The advantage of `title` in the old implementation was that it just worked for these checks (though a `title` on buttons isn't actually generally accepted to be accessible, I think).

**Review by miguelaenlle — COMMENTED** [review]
Here are some comments I had while reading the code and testing some of its functionality locally. Overall, the implementation looks great to me 👍

File: apps/prairielearn/src/pages/instructorIssues/instructorIssues.html.ts [hunk 23]
```diff
@@ -164,17 +164,17 @@ export function InstructorIssues({
...
               </a>
               <button
                 class="btn btn-med-light"
                 type="button"
-                title="Show filter help"
-                data-bs-toggle="modal"
+                data-bs-toggle="modal tooltip"
                 data-bs-target="#filterHelpModal"
-                aria-label="Show filter help"
+                data-bs-title="Filter help"
```
[the remark this answers is not in the document]
Comment (nwalters512) [reply apps/prairielearn/src/pages/instructorIssues/instructorIssues.html.ts:177]: Interesting find! FWIW, GitHub's tooltips has the same behavior. For instance, go to https://github.com/PrairieLearn/PrairieLearn/blob/master/.changeset/README.md, tab to the copy button, then mouse over the download button, and you'll see something similar:

Screenshot 2025-03-04 at 12 39 32

If we wanted to fix this, we could probably do so by hooking into Bootstrap's event system to ensure that no more than one tooltip is shown at once. We do this with popovers already.

https://github.com/PrairieLearn/PrairieLearn/blob/16255c726a79235c8fe5e7921d10758666ca97d2/apps/prairielearn/assets/scripts/behaviors/popover.ts#L88-L91

That said, I don't think this is a huge issue and probably isn't worth addressing in this PR specifically.

File: apps/prairielearn/src/components/QRCodeModal.html.ts
Comment (mylesw27) [inline apps/prairielearn/src/components/QRCodeModal.html.ts:1]: This new QR code modal looks great!

File: apps/prairielearn/src/components/CourseRequestsTable.html.ts (hunk 15 above)
Comment (mylesw27) [reply apps/prairielearn/src/components/CourseRequestsTable.html.ts:90]: I also tested this and was not able to produce any strange overflow issues.

File: apps/prairielearn/src/pages/administratorCourses/administratorCourses.html.ts [hunk 24]
```diff
@@ -139,7 +139,7 @@ export function AdministratorCourses({
                         data-bs-container="body"
                         data-bs-html="true"
                         data-bs-placement="auto"
-                        title="Really delete ${course.short_name}?"
+                        data-bs-title="Confirm deletion of ${course.short_name}"
```
Comment (mylesw27) [inline apps/prairielearn/src/pages/administratorCourses/administratorCourses.html.ts:142]: Nice update on verbiage here.

File: apps/prairielearn/assets/scripts/behaviors/tooltip.ts (hunk 18 above)
Comment (mylesw27) [reply apps/prairielearn/assets/scripts/behaviors/tooltip.ts:15]: This seems to work great. I also noticed the issue that Miguel discussed lower (in the instructor issues page) with the tooltip being always open upon modal dismissal. Not sure if that is from this code or stems from elsewhere. Either way, I don't think it's really much of an issue. This code is very clever!

File: apps/prairielearn/assets/scripts/behaviors/dropdown.ts (hunk 17 above)
Comment (mylesw27) [reply apps/prairielearn/assets/scripts/behaviors/dropdown.ts:57]: Tested and was able to tab into the popover.

**Review by mylesw27 — APPROVED** [review]
I tested locally and most everything seemed to work as expected, including testing popovers with a screenreader. There is some very nice cleanup in the code here. I only saw one minor issue and then realized Miguel had already addressed it in another comment. So, looks good to me!

File: apps/prairielearn/assets/scripts/behaviors/tooltip.ts (hunk 18 above)
Comment (nwalters512) [reply apps/prairielearn/assets/scripts/behaviors/tooltip.ts:15]: This happens because after the modal closes, we shift focus back to the last element that had focus before the modal opened (which is going to be this button) and the tooltip shows upon the tooltip receiving focus. I'd consider this to be the expected behavior, though we can probably tweak things to avoid having multiple tooltips shown at once.

File: apps/prairielearn/assets/scripts/behaviors/dropdown.ts (hunk 17 above)
Comment (nwalters512) [reply apps/prairielearn/assets/scripts/behaviors/dropdown.ts:57]: Thanks for testing!

File: apps/prairielearn/src/components/CourseRequestsTable.html.ts (hunk 15 above)
Comment (nwalters512) [reply apps/prairielearn/src/components/CourseRequestsTable.html.ts:90]: Thanks for checking 👍
